pkg/chunkenc/memchunk.go

```diff
@@ -1228,10 +1253,114 @@ func (si *bufferedIterator) moveNext() (int64, []byte, bool) {
 				continue
 			}
 			si.err = err
-			return 0, nil, false
+			return 0, nil, nil, false
 		}
 	}
-	return ts, si.buf[:lineSize], true
+
+	if si.format < chunkFormatV4 {
+		return ts, si.buf[:lineSize], nil, true
+	}
+
+	// TODO: This is pretty similar to how we read the line size, and the metadata name and value sizes
```

Comment:
I think we should do it on a separate PR since:
- The changes to extract these to a separate function are not as simple as copy-pasting the code to a separate function. How we read the timestamp and line size is slightly different from how we read the number of labels and the length of the label names and values.
- To make sure our refactoring doesn't worsen the performance, we'd need to run some benchmarks. That'll require time and this PR is blocking https://github.com/grafana/loki/pull/9702
- Sandeep will work on implementing string interning inside the chunks for the metadata labels. That'll change this `moveNext` method so I don't think it's with refactoring it until we make those changes.

I've created this task to keep track of this optimization: https://github.com/grafana/loki-private/issues/792